### PR TITLE
[WIP] Float.is_rational, Float(.5).q = 2, Float(integer).is_integer

### PR DESCRIPTION
**Pull request by smichr**
#### References to other Issues or PRs

WIP

#### Brief description of what is fixed or changed

Floats are a subset of Rationals whose denominator is a power of 2 and whose action in expressions is to moderate the precision of the result:
```python
>>> Float(0.1,5)._prec
20
>>> (10**100*Float(0.1,5))._prec
20
```
This precision only comes to play when a *result* is computed. Mathematical properties, however, are not affected by whether a number is an integer or a Float (or whether a number is any other Rational with denominator that is a power of 2): `Float(integer)*x == integer*x` and `Float(Rational(a, 2**e))*x == a/2**e*x` so `(x*y)**1.0 == x**1.0*y**1.0` and `sqrt(x) == x**.5`.

Toward this end, Float is now defined with `is_rational=True` and `is_integer=True` when the denominator of the number being represented is 1. In addition, when `x.is_rational` is tested, often another test is `...and x.q == 2`. So for this case, `Float.q=2 and for the case when `x` is an integer, `Float.q=0` (though I am not sure this is necessary; I set it to 0 so it wouldn't be confused with Integers that never have a q of 0) and for all other cases it is None. (There were only two other places where q was checked and that was in `_minpoly_cos` of `numberfields` where a check for q of 7 and 9 is made.) 

#### Other comments

I am only seeking to represent the intents and rationale of this PR in what was said above. It is not meant to be a definitive answer. And what exactly we want Float to mean -- and what it should mean -- has been an ongoing discussion spread over many issues.

#### Release Notes

NO ENTRY

### CODE REVIEW

**Comment by czgdp1807** [comment]
@smichr Thanks for making this PR. Is the issue for discussion over `Float` still needed?

**Comment by jksuom** [comment]
I think that there are many aspects of Floats that would merit from an extended discussion.
>  precision only comes to play when a result is computed.

This is the main point. The question is: How should precision affect the result?

Perhaps the simplest case is that of computing `x**-1` for a Float `x` that is represented by `p/q`. It cannot be represented by `q/p` as `p` is odd. The solution is to find an approximate representation `p'/q'`  where `q'` is a power of 2, close enough within the precision. (I skip the details here.)

The matter becomes less obvious when two Floats are added or multiplied. It seems that currently the precision of the result is determined by the *maximum* of the two precisions: https://github.com/sympy/sympy/blob/fc66c081b1f5a90654db46cd4f2612f94d7185e8/sympy/core/numbers.py#L1244-L1245
This means that when a large Float `x` of low precision and a small Float `y` of high precision are added, the result will have high precision even if `y` is smaller that the uncertainty in `x`.

**Comment by smichr** [comment]
> This is the main point. The question is: How should precision affect the result?

Is this a separate issue from this PR, however?

Regarding the issue of computation, that's an interesting point. An example:

```python
>>> n=S(849)/131072
>>> Float(n)==1/(1/Float(n))
False
```

**Comment by smichr** [comment]
> It seems that currently the precision of the result is determined by the _maximum_ of the two precisions:

I noticed that, too. I suspect that for a complex calculation it is not trivial to determine what to report for the final precision, read "propagation of errors".

**Comment by jksuom** [comment]
> Is this a separate issue from this PR

I am not quite sure that it would be totally unrelated.

File: sympy/core/numbers.py [hunk 1]
```diff
@@ -1017,12 +1017,9 @@ class Float(Number):
...
     False
     """
-    __slots__ = ['_mpf_', '_prec']
+    __slots__ = ['_mpf_', '_prec', 'q']
 
-    # A Float represents many real numbers,
-    # both rational and irrational.
-    is_rational = None
-    is_irrational = None
+    is_rational = True
```
Comment (asmeurer) [inline sympy/core/numbers.py]: I recall there being a long discussion about this at some point, and we agreed that None should be used. Probably blaming the code will lead you to the relevant issue.

**Comment by smichr** [comment]
> I am not quite sure that it would be totally unrelated.

Here is where it matters:
```python
>>> eq = Eq(x**2.0, x**2,evaluate=False)
>>> eq.subs(x,12345)
True
>>> eq.subs(x,1234567890123456789)
False
```
A Float acts like a function in a computation that truncates results of a computation. 

But this PR is not trying to address a computational issue. To point, the recently fixed `gamma(-1.0)` raises an error. The Float(-1) is equivalent in value to -1. There is no reason to delay raising an error. A given Float at a given precision is or isn't equal to an integer. The fact that `1 + 1e-20 == 1` affects Floats, too, because the precision of the Float limits what it will try to retain:
```python
>>> Float(1) + Float(1e-20) == 1
True
>>> Float(1,20) + Float(1e-20)
False
>>> (Float(1)+x).evalf(subs={x:1e-20})==1
True
```
But that is the user's issue and since we have no idea, at evaluation, of knowing what produced a given Float we can't delay evaluation because Float(1) would have been produced by any number within a range of `+/-1/2**53`. At least that's how I am inclined to see it.

File: sympy/core/numbers.py (hunk 1 above)
Comment (smichr) [reply sympy/core/numbers.py]: Perhaps issue #9636 fixed/discussed in #9670. I looked at that but think the reply [below](https://github.com/sympy/sympy/pull/16916#issuecomment-497046929) applies: there is nothing different in terms of mathematical properties between Float(1) and 1, only in how they behave in a computation.

Comment (jksuom) [reply sympy/core/numbers.py]: That was my suggestion. (And I still feel that is how it should be.)

Comment (jksuom) [reply sympy/core/numbers.py]: We already know that "equal" numbers do not necessarily behave in the same way. For example if `x` is a Float and `y` is Rational then `x == y` does not imply `x + a == y + a` for all numbers `a`.

**Comment by asmeurer** [comment]
I think it's dangerous to say that Float is rational/integer/etc. from the assumptions system point of view. The assumptions system has facts that are based on the formal mathematical rationals, integers, etc., which may be violated by floats. In particular with something like satask, any inconsistency in the facts will be detected by the solver and result in an exception.

For instance, you can see that there are issues in setting is_integer=True for floats that represent exact integers, because only integers up to the precision can be represented exactly. So `Float(10000000000000000, 15)` would not really be an integer, since it exactly the same as `Float(10000000000000001, 15)`. However, this propagates downward, since `Float(5000000000000000, 15) + Float(5000000000000000, 15) == Float(10000000000000000, 15)`, and so on. The assumptions have facts that must necessarily be violated by floats, because of the way the truncated arithmetic works. Your example where `x != 1/(1/x)` is also a good one. For rational specifically there may be issues with auto evaluation, for instance, `float**rational` always evaluates, giving a "`rational**rational`" that is never "irrational". 

I think we have no choice but to banish floats from the assumptions, at least from rational downward (real may not be a problem).

File: sympy/core/numbers.py (hunk 1 above)
Comment (smichr) [reply sympy/core/numbers.py]: Here is an example where the Float/float contains information about its history:

```
>>> 1 + 1e-20
1.0
>>> _ == 1
True
```
It's true but the fact that it's a float tells us that we *may* have been a truncated result and the expression that led us to that result may not be 1. Float behaves the same:
```python
>>> S(1)+1e-20
1.00000000000000
>>> _==1
True
```
If it is this uncertainty in history that we want to represent with Float, then we should disable `Float(1) == 1` from being True. A Float, then, is a way of using a single value to represent an interval of arbitrarily small width containing (as I think you said) and infinite number of Rationals (and irrationals?)

**Comment by smichr** [comment]
> I think we have no choice but to banish floats from the assumptions

I am starting to think (for the reasons [above](https://github.com/sympy/sympy/pull/16916#discussion_r288710351) that we should revert `Float(1) == 1`.

**Comment by jksuom** [comment]
> we should revert Float(1) == 1

I am not sure that should be done. We wanted to follow Python:
```
>>> float(1) == 1
True
```
Somehow, Python can manage that. We should accept ` Float(1) == 1` if we want to assimilate to Python as closely as possible.

**Comment by smichr** [comment]
Please see #16924

**Pull request closed without merging**